dockertest/output/dockerinfo.py

```diff
@@ -0,0 +1,100 @@
+                                                  close_fds=True)
+        self.info_string = info_string
+        self.table = self._build_table(info_string)
+
+    @staticmethod
+    def _build_table(info_string):
+        """
+        'docker info' returns a human-readable list of key: value pairs.
+        Some are indented by a space, indicating that these key/values
+        are subelements of a previous element. To wit:
+
+             Images: 3
+             Server Version: 1.12.6
+             Storage Driver: devicemapper
+              Pool Name: vg--docker-docker--pool
+              Pool Blocksize: 524.3 kB
+              ...
+             Logging Driver: journald
+
+        In this case 'Images', 'Server Version', and 'Logging Driver'
+        are simple tuples but 'Storage Driver' has both a value and
+        a set of further key/value pairs: Pool Name, Pool Blocksize, ...
+
+        We parse that and create a dict with the expected key/value
+        mapping *and* an extra: for all elements with subelements,
+        'element...' (element name plus three dots) is a dict containing
+        the subelements. E.g. x['Storage Driver...']['Pool Name'] = 'vg--etc'
+        """
+        table = {}
+        current_key = None
+        for line in info_string.splitlines():
+            # Almost every line will be Foo: Bar, but 'Insecure Registries:'
+            # is followed by a simple list of IPv4 netmasks
+            if ': ' in line or line.endswith(':'):
+                (key, value) = [e.strip() for e in line.split(':', 1)]
+            else:
+                key = line.strip()
+                value = ''
+            if line.startswith(' '):
+                if not current_key:
+                    raise SyntaxError("docker info: indented line '%s'"
+                                      " has no preceding unindented lines."
+                                      % line)
+                if current_key not in table:
+                    table[current_key] = {}
+                table[current_key][key] = value
+            else:
+                table[key] = value
+                current_key = key + '...'
+        return table
+
+    def get(self, key):
+        """
+        Each 'docker info' field has a scalar value; some also have
+        a further set of key/value pairs. Return both.
+        """
+        if key in self.table:
+            return self.table[key]
+        return None
+
+    def sub_fields(self, key):
+        """
+        Fetcher for multi-valued docker info elements. Always returns
+        a dict, even if empty.
+        """
+        sub_key = key + '...'
+        if sub_key in self.table:
+            return self.table[sub_key]
+        return {}
+
+    def __getitem__(self, key):
+        """
+        Convenience method: allows treating results as a dict
+        """
+        return self.get(key)
+
+    def __repr__(self):
+        return '{}("{}")'.format(self.__class__.__name__, self.info_string)
```

Comment:
I don't mean to be a hardass about adding API stuff.  Rather it's that in my experience, it's _far_ easier on future maintainers if it's done "flexibly" the first time, rather than try and duct-tape it later.  Consider, "later" may come when there are many more dependent callers.  I know you know, the burden of fixing library code AND all it's callers, and then testing all that, and hoping you never have to see the code again.

If instead there's at least a small level of indirection, double-buffering, and/or getters/setters type things, then it's much easier to hide future fixes w/o having them affect the callers.  Doesn't have to be perfect or cover 100%, just the big "could-change" stuff.  It also supports the rare (but tricky) cases where a test needs to sub-class and override a little part, to do something special that will probably never need to ever happen elsewhere.

This is also why there's much more strict pylint checking on ``dockertest`` over than subtests.